lib/consul/power.rb

```diff
@@ -99,10 +99,21 @@ def current=(power)
         Thread.current[ClassMethods.thread_key(self)] = power
       end
 
-      def with_power(inner_power, &block)
-        unless inner_power.is_a?(self) || inner_power.nil?
-          inner_power = new(inner_power)
+      def with_power(*args, **kwargs , &block)
+        if args.size == 1 && kwargs.empty?
+          inner_power = args.first
+          unless inner_power.is_a?(self) || inner_power.nil?
+            inner_power = new(inner_power)
+          end
+        else
+          if kwargs.empty?
+            # this branch is only required to support Ruby < 2.7.3
+            inner_power = new(*args)
+          else
+            inner_power = new(*args, **kwargs)
+          end
         end
```

Comment:
This is pretty complicated and I never liked the use of the name `inner_power` for the first argument as it does not have to be true.
We also only have two special cases, someone could provide a Power or nil, the other cases are pretty much the same (except for `kwargs` in old rubies). So how about something like this:

```
inner_power = if args.first.is_a?(self)
  args.first
elsif args.present? && args.first.nil?
  nil
elsif kwargs.empty?
  new(*args)
else
  new(*args, **kwargs)
end
```

Reply:
This has a different behavior, if someone provides a power or nil as the first argument and then provides additional rest args or keyword args.

In your implementation, the additional args would be discarded (which might be unexpected). 

In my implementation, the power/nil together with all further arguments would be passed to the initializer. I don't know, if there is any valid use case for this, but it might be slightly less unexpected than silently dropping arguments.